Bad Bot Discovery

ProgrammerDan opened the pull request:
Alright, tons of changes here. I can clean up the history if desired, but let's get some eyes on it for now. 
Adds:

* @Maxopoly 's small change discriminator, expanded to top N list

* My bounding box estimation. Might require tweaking.

* My truebot detection, which measures tile and entity lag around the player.

* Local confirmation of actual functioning on all the above, just needs tweaking and balancing.

* Tons of code fixes and cleanups; this codebase isn't as bad as some, but has some rough spots, trying to clean them up.

* Expansive debug within BotDetector, configurable on/off. Critical to determine balance, may need to expand it a bit within LagScanner to better tune what can be considered a lag source.

* Immensely expanded config to deal with the above.

ttk2 commented:
Yay, this is going to be pretty nice, if you want time to squash I can give it to you, otherwise I am eager to get a build in testing.

ProgrammerDan commented:
I put a build up for now already on civtest.

I figure i want at least one dev to greenlight my changes before any merge or squash. Although i could just make a backup and squash this...

Maxopoly commented on config.yml:
```diff
@@ -55,5 +168,5 @@ Str15: "AFKPGC version: %s, the plugin is %s"
 Str16: "AFKPGC is alive"
```
These can be removed, they were used by the old CommandHandler and the Message class, which was completly removed

Maxopoly commented on src/com/github/Kraken3/AFKPGC/BotDetector.java:
```diff
@@ -0,0 +1,306 @@
+							AFKPGC.debug("Skipping ", playerUUID, " due to reprieve or immunity");
+						}
+					} else {
+						AFKPGC.debug("Skipping ", playerUUID, " due to insufficient location data");
```
Doesn't this just cause unnecessary spam? With 2 players on the test server it's probably okay, but if we test this on the main server with ~100 players, it would send 1000 messages before even starting to do something.

ProgrammerDan replied:
Yeah, this is intentionally spammy. For tuning purposes we need *more* data. Note they are all using the AFKPGC.debug() method, which is completely turned off by the config setting ```debug: false``` which is the default. These won't be on normally, but are invaluable while configuring.

Maxopoly commented on src/com/github/Kraken3/AFKPGC/Warning.java:
```diff
@@ -0,0 +1,18 @@
+/**
+ * Lightweight warning message concerning kick nearness.  
+ */
+class Warning{
```
Is this still used anywhere after Message was removed?

ProgrammerDan replied:
Didn't even check. If you want to make a pull against my repo pulling out this dead code and the dead config options, that'd be great.

Maxopoly commented on src/com/github/Kraken3/AFKPGC/LastActivity.java:
```diff
@@ -0,0 +1,111 @@
+				newBounds.contains(lastBounds) || lastBounds.contains(newBounds),
+				newBounds.containsExcludesY(lastBounds) ||
+				lastBounds.containsExcludesY(newBounds),
+				newBounds.volume() * relaxFactor >= lastBounds.volume() ||
```
I think this is always true

ProgrammerDan replied:
It's definitely not -- observationally this is often false during testing. Since its volume it's the functional equivalent of the distance metric, just the ```Xdiff * Ydiff * Zdiff```. In this case, it's ```Xdiff * Ydiff * Zdiff * relax >= Xolddiff * Yolddiff * Zolddiff``` which is true if and only if the volumes are nearly the same. If old volume was 500 and new volume is 400, even if relax factor is 1.1 it's still 440 >= 500 which is false.

Maxopoly replied:
With argument I meant this whole section, but it didn't show both lines next to my comment

```
newBounds.volume() * relaxFactor >= lastBounds.volume() || lastBounds.volume() * relaxFactor >= newBounds.volume()
```

I mean one of the volumes is always bigger/they are the same, so one side of this OR is always true and the whole thing is true.

ProgrammerDan replied:
Yeah, I should stop coding at 2am. I see your point. 

This whole section should be:

<code>
Difference of volumes <= newBounds.volume() * relaxFactor || Difference of volumes <= lastBounds.volume() * relaxFactor
</code>

where ```relaxFactor``` is ```[0,1)```, preferably very small.

This was supposed to be a similarly test, but yeah, I screwed it up. I'll fix it. Thanks!

Maxopoly commented on src/com/github/Kraken3/AFKPGC/LastActivity.java:
```diff
@@ -0,0 +1,111 @@
+				lastBounds.containsExcludesY(newBounds),
+				newBounds.volume() * relaxFactor >= lastBounds.volume() ||
+				lastBounds.volume() * relaxFactor >= newBounds.volume(),
+				newBounds.travelSurface() * relaxFactor >= lastBounds.travelSurface() ||
```
Same for this one

ProgrammerDan replied:
Again, it's the same argument as above, just exclude the Y factor.

ProgrammerDan replied:
Same fix as above. Newfriend screwup on similarity test. Thanks for catching!

Maxopoly commented on src/com/github/Kraken3/AFKPGC/LagScanner.java:
```diff
@@ -0,0 +1,200 @@
+				for (int z = oZ - r; z <= oZ + r; z++) {
+					if (x > oX - r && x < oX + r &&
+						z > oZ - r && z < oZ + r) {
+						continue; // don't retread inner chunks of prior radius
```
Why? This filters out every chunk, which is inside the square (including the one the player is in) and instead only uses those at the square's outer border.

ProgrammerDan replied:
Read the algorithm closely. We start in the middle, scanning the chunk of the "center" location. Then we scan the next "layer" around that middle. Then the layer outside that. 

This is an "onion" function. I don't retread ground, but I also don't start in the top-left corner and scan top left to bottom right. Instead, I scan from middle out.

Maxopoly replied:
Wait yeah, I feel stupid. I've never seen something like this before, quick question completly unrelated to this and to minecraft in general: Is there any advantage to doing it this way from an algorithm view opposed to the classic row by row?

ProgrammerDan replied:
There is only an advantage is we terminate early; the "theory" is that the chunk closest to the user's location is most likely the "worst" chunk, so if we start there we can end the search quickly. We only search "outer" chunks if the chunks right around the player aren't enough to convict him.

Maxopoly commented on config.yml:
```diff
@@ -1,21 +1,134 @@
+
+lag_scanner:
+  # Length of time to cache individual chunk results
+  cache_timeout: 900000
```
You should probably add that this is in ms, because some other timings are in ticks. Also I think we can safely increase this to half an hour, it's almost impossible to cause bigger differences without creative mode in less than that.

ProgrammerDan replied:
Yeah, definitely. For testing I need it to basically not cache, but definitely on the main server this can be more generous.

Maxopoly commented on src/com/github/Kraken3/AFKPGC/LagScanner.java:
```diff
@@ -0,0 +1,200 @@
+					}
+					if (totalCost >= lagSourceThreshold) {
+						AFKPGC.debug("Chunk ", chunkId, " passed source threshold with tiles and entities");
+						break; // if we cross the threshold, don't keep adding.
```
Dont we want to know how bad it really is? If we do, we could easily determine, who are the hardest offenders, right now it will just log the threshhold when ever it finds someone having a ton of entities.

ProgrammerDan replied:
Well, I'm conflicted here. Scanning a full area around a player might take 25-30ms, so not much time, and yes, this shouldn't be run very often. From a "tracking" perspective we shouldn't break out early. From an algorithmic efficiency purpose, we must. Perhaps I could add a config option that determines if it "breaks early" or records the full damage.

Maxopoly commented on src/com/github/Kraken3/AFKPGC/BotDetector.java:
```diff
@@ -0,0 +1,306 @@
+						truebot += (bounds.getNearlyContainedExcludeY() ? 
+								boundsConfig.getNearlyContainedExcludeY() : 0.0);
+
+						if (truebot >= boundsConfig.getThreshold()) {
```
Shouldn't the total size of the box also be taken into account here? It's possible to slip through those checks with different boxes, which are each smaller than 10 blocks. Also how about using online time ?

ProgrammerDan replied:
I'm not sure you're understanding how the bounding box works?  10 Blocks has nothing whatsoever to do with this. The bounding box is the full extent of the Locations as captured by previous invocations of BotDetector -- it encompasses the *maximum* area marked out by all the locations. This could be a volume encompassing hundreds or thousands of blocks. 

Online time doesn't really matter, it doesn't take too long for a lag machine to spin up and ruin things for everyone :).

Maxopoly commented on config.yml:
```diff
@@ -1,21 +1,134 @@
+    PISTON_STICKY_BASE: 2
+    POWERED_RAIL: 1
+    REDSTONE_COMPARATOR: 1
+    REDSTONE_COMPARATOR_OFF: 1
```
There are huge circuits, which are rarely used, but would get players kicked just for being close to them, even though  it wasn't doing anything. I dont think we should take turned off redstone components into account here.

ProgrammerDan replied:
It would need to be a truly massive circuit. I have these numbers set pretty low, for that reason. We can always change these values to 0 if we find that it's targetting redstoners incorrectly.

Maxopoly replied:
Because of the cart lag and fully powered carts stopping on the tracks, it got common practice to use powered rails only for any sort of cart collection system. I know enough places that have 1k+ powered rails, those machines usually aren't used that often, but right now it would ban people just for walking close to those monsters. Do powered rails cause lag if they are not powering any carts?
Also what brought up the thought for me was, that for example Thule has about 1k+ comparators below the city, which are intended to be used once every few months. People would pretty much get banned there all the time for standing in the middle of the town, while actually no redstone is powered and no lag is being caused.  
Independent from this, I think that many people aren't really aware how much lag they cause with tile entities (maybe that includes what I just tried to defend) and when we release this, we should tell people to reduce the amount of tile entities around them to reduce their chance of being kicked, this might bring a performance improvement on its own.

ProgrammerDan replied:
I think the messaging is the key. Large scale redstone mechanism can definitely cause lag as they are updated every tick. Even when unpowered, they still get called -- I just expect the update functions return quickly once its determined they are unpowered.

However, we're going to have this at like 17.5 maybe 16 before people start getting kicked. We can up the reprieve once we tune things, so even if someone in Thule got kicked, if it didn't improve tick they'd had a 3 minute sit out and be undisturbed for the next several hours (once tuned).

ProgrammerDan replied:
Should have said -- I'll have to look at powered rails. We might be able to 0 those safely, I don't have a good idea how much "per unit lag" they cause. I don't think they get updated every tick, although they are a redstone mechanism so they might ... (just to check if powered, for instance).

Maxopoly commented on src/com/github/Kraken3/AFKPGC/BotDetector.java:
```diff
@@ -0,0 +1,306 @@
+						AFKPGC.debug("Player ", curSuspect.getUUID(), " Not enough bounds data, skip for now");
+					}
+				}
+			} else {
```
It's still possible to be a cause of major lag, but to avoid getting caught by simply moving around in clever patterns (or just travelling through furnaceland 1.0). What do you think of including something here, that if noone was found for X rounds (or the tick goes below a lower very critical value), it just goes full hardcore mode and scans everyone until it finds someone with a ton of tile entities? This might take up a lot of resources for a minute or two and make it even worse, but it would at least find the right people, if needed.

ProgrammerDan replied:
Well, clever patterns won't trick the bounding box unless they are specifically designed to thwart BotDetector. Given the timing issue and that the latest commit of BotDetector uses variable invocation timing, it would be very hard to forever fool the detector.

That said, I totally agree that we need a HAL mode that just starts scanning and ignores other "nice" things.

Maxopoly commented on src/com/github/Kraken3/AFKPGC/BotDetector.java:
```diff
@@ -0,0 +1,306 @@
+					   				p.kickPlayer(leBan.getReason());
+								}
+								bannedPlayers.add(lastRoundSuspect.getName());
+								addToBanfile(lastRoundSuspect.getName());
```
We don't need to add him to the banfile, if longbans wasn't even enabled, that should be in the if-clause

ProgrammerDan replied:
It ... is in the if clause? I'm not sure what you're referring to here, if longbans is off no-one gets added to the ban file.

Maxopoly replied:
Nevermind I read the brackets wrong, sorry.

Maxopoly commented on src/com/github/Kraken3/AFKPGC/BotDetector.java:
```diff
@@ -0,0 +1,306 @@
+						/* This can be relatively sensitive, because it will only ban players for 
+						 * a longer period of time, if it catches them twice here */
+						Date currentDate = new Date();
+						if (suspectedBotters.contains(lastRoundSuspect.getName())) {
```
Can you invoke methods on offline players ? As far as I understood it, it throws exceptions if you try to do that, that was why I stored the name and some other stuff additionally.

ProgrammerDan replied:
This doesn't invoke methods on offline players? LastRoundSuspect is an object of type Suspect, a special container. I just did in OOD what you did with a bunch of variables. I just pulled them all together into a Suspect object; still no invocations of Player methods when offline :).

Maxopoly replied:
Ah, I just had in mind that lastRoundSuspect was a Player, from when I was writing it and didn't see that you changed it, my bad.

Maxopoly commented on src/com/github/Kraken3/AFKPGC/BotDetector.java:
```diff
@@ -0,0 +1,306 @@
+			lastRoundTPS = currentTPS;
+			lastRoundSuspect = thisRoundSuspect;
+		} else { // TPS is high enough
+			// TODO: Evaluate if this makes sense. We should probably see if TPS stays good for a while
```
We could just keep the players causing lag locked away for 30 minutes after the tick is good again, that sounds reasonable imo.

ProgrammerDan replied:
Yeah, agreed. Maybe make it a function of BotDetector invocations.

Maxopoly commented on config.yml:
```diff
@@ -1,21 +1,134 @@
+  # Summation threshold of entity and tile costs for lag flag
+  lag_threshold: 750
+  # Per tickable block cost 
+  tick_block:
```
What about portal blocks?

ProgrammerDan replied:
Find me the Tile entity, could add them, although I didn't because I think it would unfairly target simple farms. They are updated via the BUD mechanism.. I don't think they are updated every tick.

Maxopoly commented:
Overall it looks very good, it just needs some finetuning. Do want any getter/setter commands for any of your new config values for easier debugging? I could just quickly add those to take some work of you.

ProgrammerDan commented:
Since you added the reload config option it's basically unnecessary to have additional getters/setters -- you can just modify the config and reload. Pulling out the dead cruft would be useful, though! Also, add a command that will force an immediate invocation of the BotDetector, safely ... not sure how to do it yet (would need to add some new methods) but it would be useful for many reasons, especially in low TPS where the detector rarely runs, even with the latest commit's auto-balancing feature.

ttk2 commented:
@ProgrammerDan  how is this coming?

ProgrammerDan commented:
Had a great testing session last night, I've addressed some of the issues found but not all. Working with Max on fixing out some other flaws/issues. 

Evaluating adding a "HAM" or "HAL" mode that when TPS gets below a certain threshold (12 perhaps) skips the "nice guy" routine of checking movement patterns and just starts scanning chunks and kicking.

ProgrammerDan commented:
So @Maxopoly  I added a new command to force an invocation of the BotDetector. I hopefully fixed the logic on the volume/surface area bounds -- can you give it a lookover and see if my logic/code is sound? 

I'm still on the fence about computing all the chunks all the time. I think I might for one simple reason:

Kicking people doesn't immediately restore TPS.

So, the short ban is proving quite ineffective. By the time the ban is up, TPS typically hasn't improved b/c none of the "bad chunks" have unloaded. We either need to (1) force unload the chunks that the player was in, or (2) wait longer between each BotDetector invocation. If we go with (2), however, we have to "live with" bad TPS for longer, as it will take a considerable time to find and eliminate all the lag sources. Additionally, it could be foiled by the botter simply logging in and out periodically (I need to review LastActivity and ensure it is more persistent, it should not be released, etc.). 

Consequently, I want to set a few thresholds. If you're just "barely" over the lag threshold, kick you short and test. If you're far over the threshold (monster lag machine), long ban you immediately. Thoughts?

On a good note, my auto-adaptive scheduling of the next BotDetector is working *great*, in low TPS BotDetector runs more often to "compensate", this is proving to be very good.

I also have some good data on "light" TPS impact -- looks like a normal user doing normal things will measure about 150-250 on my "lagscanner". So we can probably set our first kick threshold in the 400-500 range; and if we go with the "kill with fire" technique for large lag machines, I think that threshold is probably at about 1k. (read: 200 hoppers, or blend of Tick drainers along those lines should be instagib).

Maxopoly commented:
Your changes look good and the adaptive scheduling is a really neat idea.
I also thought about making longer bans based on tps increase and I dont think we should reduce the frequency of the detector, we can just test it as is and see the results. Does spigot evensupport forcibly unloading specific chunks? 

The insta-ban thresh hold sounds like a good idea, but it should probably be a multiplier of the normal thresh hold instead of a flat value.

Couldnt we just add a config option to disable short bans, set that to true and let it debug for like an hour on the actual civcraft server today or tomorrow? I feel like we need some actual data to work with for a balanced config.

ProgrammerDan commented:
That's a great idea actually. Ok, so let me sum up the changes for tonight, assuming no issues in testing we could probably stage it in "softcore mode" for Friday.

* Get Config to a safe "softcore" version to run on main Civcraft. This includes:
    * Good caching length (10x current).
    * Good reprieve length (2x current)
    * Reduce "normal" threshold based on observations in Civtest
* Add option to disable BotDetector kicking/banning entirely (softcore, call it "Observation Mode"). Note this would not impact standard AFK kicker.
* Add option for instagib threshold as a multiple of normal threshold. 
* Add second option to force LagScanner to scan all chunks always
* Alter LagScanner to respect all scan, if off alter to scan up to instagib threshold
* Clean up some of the more horribly spammy debug statements that don't contribute useful information
    * Suspect list management; should only output the suspects
    * Chunk scanning; one debug line per chunk scan STRICT
    * General scan for others

Finally research:

* Can we force unload chunks? Civtest has chunk-gc on but many chunks remain loaded even with no-one online, causing TPS to rest at 17. With no players. Investigating timings now but persistently loaded chunks appear to be source.

Maxopoly commented:
Me and git are in disagree about some stuff, but I'm working on it.

ProgrammerDan commented:
Oh wow, thanks. I didn't expect you to work on any of that list, but that's great. If anything is unclear just hit me up. Can't do the code atm but can reply here :)

jjj5311 commented:
this is pretty damn cool

ProgrammerDan commented:
@Maxopoly  Ok so there's a command -- Chunk.unload(save, safe); -- if you grab the location of the player *just* before kicking him, you could "safely" attempt to force the unload of all chunks around that player. Just use a similar technique to the lag scanner, but once you have the chunk call .unload(true, true) on it, to save the chunk but only unload if no players are nearby (be "safe").

I did forget one item on my list:

* Catching multiple botters in the same machine. I have a code fragment to kick players within a near-radius of a giant lag source; I need to implement that and effectively "ban" all the near-players as a single "lag suspect". Will require some recoding so don't worry about it for now, I'll deal with it as an improvement, but sadly botters could defeat us at the moment simply by putting two accounts in every machine, and I have no intention of preserving that weakness.

@jjj5311 Thanks :D hope it works, too.

ProgrammerDan commented:
* Also, add command to force clear the reprieve list.

ProgrammerDan commented:
@Maxopoly Were you able to get any changes working? going to do a few things real quick, hope it doesn't mess you up :(.

Maxopoly commented:
Nah I spent 3 hours trying to understand git and get it to do what I want it to do and 2 minutes coding in an observation mode. (And actually just got it working)

Maxopoly commented:
I can do the config and option stuff, can you just do the chunk unloading?

ProgrammerDan commented:
Just adding a few commands real quick, but yeah I'll handle the chunk unloading. I'm going to make it part of LagScanner, basically just try to unload "bad" chunks (some configurable fraction of low threshold) b/c unloading chunks is kind of expensive (the saving is).

ProgrammerDan commented:
Ok just pushed @Maxopoly -- just a few quick commands. Should be safe to merge into your changes or rebase. Thanks! I'll hit the other bits later tonight.

Maxopoly commented:
Okay, I made a pullrequest to your repo for all the config stuff, so the only things left are the forced chunk unloading, rebalancing the config a bit and the banning of nearby players.

Chunk unloading is yours and the config rebalancing should also be done by you, because you saw the actual logs from civtest (though the main part of balancing will still come from the test on the main server)

The banning of nearby players is a bit difficult, because if we determine someone a lag source after kicking him, because the tick improved, kicking players around him wouldn't make sense, because those chunks were either never unloaded and we have a false positive or those players are not loading the chunks in question somehow.
While it makes sense to ban the players around chunks, which were determined an extreme lag source, it gets difficult if it comes to larger server events, where 25 % of the server population or more are at one location. In this case it would be okay for that place to consume a fourth of the current resources and we shouldn't just ban all those people.

Nevertheless I think we can test this tomorrow and it makes me really happy that we are finally at a point where we can do that and get very close to fully releasing this.

ProgrammerDan commented:
Thanks, I'll check that out shortly. 

Yeah, I want nearby banning as a tool, but I'm talking *nearby* -- e.g. within 24 blocks or something. Basically on top of. Perhaps with a slightly different ban message -- "you were found in chunks responsible for generating significant lag, if you're not a bot please leave this area when your ban lifts in X seconds". 

Basically, if they are practically in the same chunk as the player we've identified as a serious lag source, they've got to go. Otherwise, the TPS will not improve, and we will fail to identify the player we did kick as a source of lag.

Large server events are fine, provided they don't involve spamming tons of hoppers or pistons :D Those won't even be a blip on the lagscanner's radar. I also think the associative ban should only be if the lagsource amount is above the "horrible" threshold. 

Does that make sense at all?

Also, equally excited. This bad TPS has been depressing the server for too long, everyone's attitude is mad salty as a result.

Maxopoly commented:
It makes sense, but I think we should just base it on testing.

ProgrammerDan commented:
My comments are actually based on testing. Sintralin was co-habitating a lag monster with Ariaxis the other night (might have the names wrong) -- Ariaxis got kicked, lag monster didn't unload b/c of sintralin, and consequently both were granted reprieve.

Maxopoly commented:
Ah I see, yeah that needs to be adressed.

Maxopoly commented:
Also I got another thought earlier, once this goes live, the rule against unreasonable lag machines needs to be inforced a harder, because people could for example just place 1k hoppers around their vault to keep anyone from breaking it, because anyone going there for longer would get banned. Also grief through placing tile entities near your opponents cities could become a thing, should intervention against this come from an admin level? Also many people I talked to said, that at this level we should not keep the code of the plugin and especially the config public, what do you think of that?

ProgrammerDan commented:
If the plugin is public or not, we're still be in a war of sorts with botters. Largely this exists to take the edge of server lag *in general* and help identify the people who are majority responsible for the sources of lag. That equips the administration if they choose to take more "serious" action in reigning in lag. We're not going to solve it here, not entirely, but we can definitely make a difference and that's what is needed right now. 

I don't think making the code secret will really change anything 'cept long-tail the struggle a bit.

ttk2 commented:
@Maxopoly  that vault idea... why do I never think of these things.

ProgrammerDan commented:
I mean it's literally what ShadedJon did with FurnaceLand, sans the actual vault. The nice thing is now we'll know where these lag monsters are without the admins trying to fly around everywhere.

ProgrammerDan commented:
@ttk2 can you PM me the current AFK config for main? I'll combine it and the new options and reply with a tailor config for staging tomorrow. We can also discuss which options to enable/disable.

ProgrammerDan commented:
@Maxopoly When you get a chance, give my changes a once over. I think I've got the major issues ironed, and quick testing on Civtest demonstrates its working well. Long bans is not well tested ATM, so before we turn it on we should do some more testing of it. Associative kickbanning, the same. Not an issue for initial "data collection" run.

The only thought I had was to give TTK a break and add turning on/off the debug output as a console command. Reactions?

Maxopoly commented:
@ProgrammerDan  I think we should reconsider the whole concept of suspects. When I implemented that, I wanted to reduce the amount of false positives, because at that point the lag detection was horribly bad compared to what we have now. If we assume we manage to create a balanced config, there is no reason for us to test whether someone is causing lag with the entities he loads, we already know he is, if he exceeds the threshold. We could instead just message him and say "Hey, get away from there" and if he doesn't, he gets a long ban. This would also be a lot fairer from a player perspective, because unless they exceed the threshold for extreme lag, they wouldn't just get a ban suddenly and we could reduce the amount of false positives massively, if some still exist.

ProgrammerDan commented:
I'm not in opposition to this idea, but let's get this version running and pulling testing data on before we refactor again.

That said, wanna take my branch, fork it, and run with your idea a bit? At least now our toolchest is bigger, we can definitely explore with more confidence how to use it effectively.

Maxopoly commented:
Yeah, that's a good idea.

ProgrammerDan commented:
@ttk2 Added that debug mode toggle I mentioned. Tested on Civtest, all is well, as is the config I sent you. Lemme know if a build isn't available on the build server for some reason.

ProgrammerDan commented:
@ttk2 went ahead and squashed 44 commits down to 8, good to merge.
